# pRIC

#### Description

This is the code repository for analyzing pRIC-seq data in the project entitled "Global atlas of enhancer-promoter interactome in cotton genome revealed by profiling RNA-RNA spatial interactions".

#### Developers

+ Miaomiao Wen
+ Yijie Wang


#### Usage

* Workflow of data analysis (Snakemake files and Python scripts):
  + pRIC/: Data processing of pRIC-seq.
  + ChIP/: The data analysis of ChIP-seq.
  + DNase/: The data analysis of DNase-seq.
  + GRO/: The data analysis of GRO-cap.
  + AD_RNA/: The data analysis of RNA-seq in tetraploid.
  + VIGS_RNA/: RNA-seq data analysis of VIGS-mediated gene silencing.
  + GWAS/: Integrated analysis using published GWAS data.
  + figures/: R codes for generating all figures presented in the manuscript.

* Chimeric Read Extractor for pRIC-seq data (pRIC/scripts/chimeric)

  + Extracts chimeric reads from STAR alignment outputs (fetch_chimeric_from_STAR_aligned.py).
  + Extracts chimeric reads from BWA-MEM alignment outputs (fetch_chimeric_from_BWA_aligned.py).

Script Usage:

    fetch_chimeric_from_STAR_aligned.py

     -b, --bam : BAM file generated by STAR aligner
     -c, --chimeric : SAM file generated using STAR aligner with parameter '--chimOutType SeparateSAMold'
     -R1, --read1 : Paired-end sequencing R1 file (R1.fastq.gz)
     -R2, --read2 : Paired-end sequencing R2 file (R2.fastq.gz)
     --known-js : a Bed6 formatted file containing known splice sites
     --rule: strand-specific library construction mode, default: '1+-, 1-+, 2++, 2--'
     -o, --output : output file containing chimeric read information

    fetch_chimeric_from_BWA_aligned.py: 

    -b, --bam : BAM file generated by BWA aligner
    -R, --read : Single-end sequencing file (*.fastq.gz)
    --known-js : a Bed6 formatted file containing known splice sites
    -o, --output : output file containing chimeric read information

